lib/ecto/changeset.ex

```diff
@@ -1349,6 +1349,99 @@ defmodule Ecto.Changeset do
+
+  This is not reliable because conflicting records could be inserted by another
+  process after this validation runs. Use a constraint for a reliable check.
+
+  However, in most cases when there are conflicts, they will already exist by
+  the time of validation. Checking for them in the validation phase may allow
+  users to fix errors like "username is already taken" at the same time as
+  errors like "email can't be blank", instead of fixing each constraint error
+  the database raises, one at a time, after the validations pass.
+
+  To use this, pass a function that finds existing records with which this
+  changeset would conflict, according to a unique or exclusion constraint. If
+  the function returns any an empty enumerable, there are no conflicts and the
+  validation passes. If the function returns an enumerable containing one
+  existing record, but the changeset is an update to that record (that is, the
+  changeset and the existing record are for the same schema and have the same
+  primary key), the validation passes. Otherwise, the changeset would create a
+  conflict and the validation fails.
+
+  The `field_name` is where the error message (if any) will be placed.
+
+  ## Options
+
+    * `:message` - the error message on failure; defaults to "has already been taken"
+
+  ## Examples
+
+      validate_no_conflicts_unreliably(changeset, :title, &find_posts_with_title/1)
+      validate_no_conflicts_unreliably(changeset, :title, &find_posts_with_title/1, message: "not original")
+      validate_no_conflicts_unreliably(changeset, :base, &find_overlapping_rentals/1, message: "an existing rental overlaps these dates")
+
+  A finder function might look like:
+
+  defp find_posts_with_title(changeset) do
+    title = get_field(changeset, :title)
+    if is_nil(title) do
+      # no title, so there can be no conflicting records
+      []
+    else
+      MyApp.Repo.all(from p in Post, where: e.title == ^title)
+    end
+  end
+  """
+  @spec validate_no_conflicts_unreliably(t, atom, (t -> list), String.t) :: t
+  def validate_no_conflicts_unreliably(changeset, field_name, conflict_finder_func, opts \\ []) do
+    conflicts = changeset
+    |> conflict_finder_func.()
+    |> Enum.reject(fn (apparent_conflict) ->
+      would_update?(changeset, apparent_conflict)
+    end)
+
+    if Enum.any?(conflicts) do
+      message = message(opts, "has already been taken")
+      add_error(
+                changeset,
+                field_name,
+                message,
+                [validation: :validate_no_conflicts_unreliably]
+              )
+    else
+      changeset
+    end
+  end
+
+  defp would_update?(changeset, record) do
+    source(record) == source(changeset) \
+      && primary_key(record) == primary_key(changeset)
+  end
+
+  defp source(changeset = %Changeset{}) do
+    changeset.data.__struct__.__schema__(:source)
+  end
+
+  defp source(record) do
```

Comment:
Maybe the versions of `source/1` and `primary_key/1` that take a record should belong to a different module?